### PR TITLE
NEXUS-8293 - Implement framework bits and Nuget default repos

### PR DESCRIPTION
**Pull request by kellyrob99**
- interface and first impl to back default repos

### CODE REVIEW

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/manager/RepositoryManagerImpl.java [hunk 1]
```diff
@@ -146,8 +150,14 @@ private void untrack(final Repository repository) {
...
-      log.debug("No repositories configured");
-      return;
+      log.debug("No repositories configured, installing defaults");
+      for (InitialRepositoryConfiguration configProvider : initialRepositoryConfigurations) {
+        for (Configuration configuration : configProvider.getRepositoryConfigurations()) {
+          log.debug("Provisioning default repository: {}", configuration);
+          store.create(configuration);
+        }
+      }
+      configurations = store.list();
```
Comment (jdillon) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/manager/RepositoryManagerImpl.java]: configurations could still be empty if no InitialRepositoryConfiguration or they returned empty lists.  Edge case but that would then show odd logging below of "Restoring 0 repositories" etc

File: plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetInitialRepositoryConfiguration.groovy [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import com.sonatype.nexus.repository.nuget.internal.proxy.NugetProxyRecipe
+
+import org.sonatype.nexus.repository.config.Configuration
+import org.sonatype.nexus.repository.manager.InitialRepositoryConfiguration
+
+/**
+ * Provide default hosted and proxy repositories for NuGet.
+ * @since 3.0
+ */
+@Named
```
Comment (jdillon) [inline plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetInitialRepositoryConfiguration.groovy]: probabably should be javax.inject.Singleton

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/manager/InitialRepositoryConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.repository.manager;
+
+import java.util.List;
+
+import org.sonatype.nexus.repository.config.Configuration;
+
+/**
+ * @since 3.0
```
Comment (jdillon) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/manager/InitialRepositoryConfiguration.java]: describe please

File: plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetInitialRepositoryConfiguration.groovy (hunk 2 above)
Comment (kellyrob99) [inline plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetInitialRepositoryConfiguration.groovy]: done

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/manager/InitialRepositoryConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+import java.util.List;
+
+import org.sonatype.nexus.repository.config.Configuration;
+
+/**
+ * Provides configurations for repositories that should be provisioned by default on first system startup.
+ * @since 3.0
+ */
+public interface InitialRepositoryConfiguration
```
Comment (adreghiciu) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/manager/InitialRepositoryConfiguration.java]: I would name it: DefaultRepositoriesContributor

**Comment by adreghiciu** [comment]
+1

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/manager/RepositoryManagerImpl.java [hunk 5]
```diff
@@ -147,7 +157,23 @@ protected void doStart() throws Exception {
     List<Configuration> configurations = store.list();
     if (configurations.isEmpty()) {
       log.debug("No repositories configured");
-      return;
+      if (Boolean.valueOf(System.getProperty(SKIP_DEFAULT_REPOSITORIES, "false"))) {
+        log.debug("Skipping provisioning of default repositories due to '" + SKIP_DEFAULT_REPOSITORIES + "' property");
```
Comment (jdillon) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/manager/RepositoryManagerImpl.java]: string concat... bad

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/manager/RepositoryManagerImpl.java [hunk 6]
```diff
@@ -147,7 +157,23 @@ protected void doStart() throws Exception {
...
+        return;
+      } else {
```
Comment (jdillon) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/manager/RepositoryManagerImpl.java]: formatting :-P

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/manager/RepositoryManagerImpl.java [hunk 7]
```diff
@@ -147,7 +157,23 @@ protected void doStart() throws Exception {
...
+        log.debug("Provisioning default repositories");
+        for (DefaultRepositoriesContributor configProvider : defaultRepositoriesContributors) {
+          for (Configuration configuration : configProvider.getRepositoryConfigurations()) {
+            log.debug("Provisioning default repository: {}", configuration);
+            store.create(configuration);
+          }
+        }
+        configurations = store.list();
+        if(configurations.isEmpty()) {
```
Comment (jdillon) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/manager/RepositoryManagerImpl.java]: formatting :-P

**Comment by jdillon** [comment]
+1 few minor bits to tidy though, please do that so I don't have to :-P

**Pull request merged**